pyartcd/pyartcd/pipelines/build_rhcos.py

```diff
@@ -0,0 +1,191 @@
+
+        return base64.b64decode(secret.model.data.token).decode('utf-8')
+
+    @staticmethod
+    def build_parameters(build: Dict[str, List[Dict]]) -> Dict:
+        """Parse the build parameters from the build actions into a more usable dict"""
+        parameters = next((  # first action that has parameters
+            action["parameters"]
+            for action in build["actions"]
+            if "parameters" in action
+        ), [])
+        return {p["name"]: p["value"] for p in parameters}
+
+    @staticmethod
+    def build_url(job: str, number: int) -> str:
+        return f"{JENKINS_BASE_URL}/job/{job}/{number}/"
+
+    def query_existing_builds(self) -> List[Dict]:
+        """Check if there are any existing builds for the given version. Returns builds in progress."""
+        builds = []
+        for job in ("build", "build-arch", "release"):
+            builds.extend(
+                dict(**b, job=job, parameters=self.build_parameters(b), url=self.build_url(job, b["number"]))
+                for b in self.request_session.get(
+                    f"{JENKINS_BASE_URL}/job/{job}/api/json?tree=builds[number,description,result,actions[parameters[name,value]]]",
+                ).json()["builds"]
+                if b["result"] is None  # build is still running when it has no status
+            )
+
+        return [b for b in builds if b["parameters"].get("STREAM") == self.version]
+
+    def start_build(self):
+        """Start a new build for the given version"""
+        # determine parameters
+        params = dict(STREAM=self.version, EARLY_ARCH_JOBS="false")
+        if self.build_always:
+            params["FORCE"] = "true"
+
+        # start the build
+        self.request_session.post(
+            f"{JENKINS_BASE_URL}/job/build/buildWithParameters",
+            data=params,
+        )
+
+        # wait for a related build to begin
+        initial_builds = []
+        while not initial_builds:  # may take a few seconds for the build to start
+            time.sleep(1)
+            initial_builds = self.query_existing_builds()
+
+        return initial_builds
+
+    def build_result(self, job, number):
+        """Query the status of a known build"""
+        return next((  # expecting exactly one result
+            dict(
+                url=self.build_url(job, b["number"]),
+                description=b["description"] or "[no description yet]",
+                result=b["result"],
+            ) for b in self.request_session.get(
+                f"{JENKINS_BASE_URL}/job/{job}/api/json?tree=builds[number,description,result]"
+            ).json()["builds"]
+            if b["number"] == number
+        ), None)
+
+    def wait_for_builds(self):
+        """Wait for all builds for this version to complete, and give status updates on stderr"""
+        builds_seen: Dict[Tuple, str] = {}
+        completed_builds: Dict[Tuple, Dict] = {}
+        while True:
```

Comment:
likewise here... if the pipeline isn't complete after maybe 6 hours, chances are something is wrong.